<a href="https://colab.research.google.com/github/Olegman11/11/blob/main/F7_Practical_task.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [32]:
# turning on pandas library
import pandas as pd
import numpy as np # import numpy

In [33]:
# connecting a dataset
Data_health_risk = pd.read_csv('/content/F7_health_risk_data_for_analysis.csv') # uploading data in pandas
Data_health_risk

,Age,SystolicBP,DiastolicBP,BS,BodyTemp,HeartRate,RiskLevel
0,52.0,NaN,Eighty,10.0,98.0,76.0,high_risk
1,45.0,160.0,70,6.7,98.0,82.0,high risk
2,33.0,90.0,100,6.2,99.0,74.0,high_risk
3,NaN,120.0,110,14.7,36.6,96.0,1
4,52.0,160.0,100,13.5,NaN,80.0,0
...,...,...,...,...,...,...,...
146,40.0,NaN,100,6.1,NaN,83.0,low risk
147,NaN,90.0,70,NaN,37.2,90.0,low risk
148,45.0,120.0,80,6.5,37.2,65.0,NaN
149,40.0,120.0,110,7.1,99.0,61.0,LOW


In [34]:
# copy dataset in case
data = Data_health_risk.copy()
data

,Age,SystolicBP,DiastolicBP,BS,BodyTemp,HeartRate,RiskLevel
0,52.0,NaN,Eighty,10.0,98.0,76.0,high_risk
1,45.0,160.0,70,6.7,98.0,82.0,high risk
2,33.0,90.0,100,6.2,99.0,74.0,high_risk
3,NaN,120.0,110,14.7,36.6,96.0,1
4,52.0,160.0,100,13.5,NaN,80.0,0
...,...,...,...,...,...,...,...
146,40.0,NaN,100,6.1,NaN,83.0,low risk
147,NaN,90.0,70,NaN,37.2,90.0,low risk
148,45.0,120.0,80,6.5,37.2,65.0,NaN
149,40.0,120.0,110,7.1,99.0,61.0,LOW


In [35]:
# info about data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151 entries, 0 to 150
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Age          127 non-null    float64
 1   SystolicBP   124 non-null    float64
 2   DiastolicBP  143 non-null    object 
 3   BS           143 non-null    float64
 4   BodyTemp     143 non-null    float64
 5   HeartRate    143 non-null    float64
 6   RiskLevel    143 non-null    object 
dtypes: float64(5), object(2)
memory usage: 8.4+ KB


In [36]:
# checking names of columns for errors, spases and so on
print(data.columns)

Index(['Age', 'SystolicBP', 'DiastolicBP', 'BS', 'BodyTemp', 'HeartRate',
       'RiskLevel'],
      dtype='object')


In [37]:
data.shape

(151, 7)

In [38]:
data.describe()

,Age,SystolicBP,BS,BodyTemp,HeartRate
count,127.000000,124.000000,143.000000,143.000000,143.000000
mean,38.251969,123.951613,10.258042,76.121678,80.769231
std,15.450879,21.444523,2.860745,29.797954,18.143672
min,18.000000,90.000000,0.100000,36.600000,60.000000
25%,29.000000,110.000000,7.500000,37.200000,70.000000
50%,35.000000,120.000000,10.300000,97.500000,81.000000
75%,45.000000,140.000000,12.800000,99.000000,89.000000
max,150.000000,160.000000,14.900000,100.000000,250.000000


In [39]:
# to take a look 10 random rows
data.sample(10)

,Age,SystolicBP,DiastolicBP,BS,BodyTemp,HeartRate,RiskLevel
103,52.0,120.0,90,11.2,98.0,98.0,LOW
41,35.0,140.0,80,13.7,100.0,90.0,high risk
14,60.0,130.0,90,14.6,97.5,89.0,high risk
94,52.0,90.0,80,8.9,37.2,82.0,LOW
99,33.0,130.0,NaN,6.8,98.0,84.0,NaN
67,33.0,90.0,90,6.2,100.0,93.0,1
134,35.0,140.0,100,13.3,NaN,63.0,high_risk
64,35.0,90.0,110,6.6,36.6,NaN,1
50,29.0,90.0,Sixty,9.5,97.5,91.0,high_risk
44,NaN,140.0,110,6.6,100.0,67.0,0


During the initial data analysis, I identified several data quality issues that need to be addressed before moving to the next stage of analysis.

First, the dataset contains 151 rows and 7 columns, but several columns have missing values. For example, the Age column contains only 127 non-null values, and SystolicBP contains 124 non-null values, while other columns contain 143 non-null values. This suggests that there are missing records, and I need to investigate whether these values were lost during data collection or data import.

Second, I found inconsistent data types. The DiastolicBP column is stored as an object instead of a numeric type. After reviewing the data, I noticed non-numeric values such as "Eighty", which should be converted into a proper numeric format.

Third, I found inconsistent category labels in the RiskLevel column. For example, the same categories appear in different formats, such as "high_risk", "high risk", "LOW", and numeric values like 0 and 1. These values need to be standardised to ensure consistency during analysis.

I also noticed possible outliers or unusual values in some numerical columns. For example, the BS column contains values with a very wide range, such as 0.1 and 14. I need to understand the meaning and unit of this variable before deciding whether these values are valid or represent data entry errors.

In [40]:
#checking RiskLevel for unique values
data['RiskLevel'].unique()

array(['high_risk', 'high risk', '1', '0', 'low risk', 'mid risk', 'LOW',
       nan], dtype=object)

In [41]:
# replacement the incorrect data for 'High Risk' and 'Low Risk' columns
data['RiskLevel'] = data['RiskLevel'].replace(['high_risk', 'high risk', '1'], 'High Risk')
data['RiskLevel'] = data['RiskLevel'].replace(['low risk', '0', 'LOW'], 'Low Risk')

In [122]:
# drop the row with NaN in RiskLevel column
data.dropna(subset=['RiskLevel'], inplace=True)

In [124]:
data

,Age,SystolicBP,DiastolicBP,BS,BodyTemp,HeartRate,RiskLevel
0,52,120,80,10.0,36.7,76.0,High Risk
1,45,160,70,6.7,36.7,82.0,High Risk
2,33,90,100,6.2,37.2,74.0,High Risk
3,37,120,110,14.7,36.6,96.0,High Risk
4,52,160,100,13.5,36.9,80.0,Low Risk
...,...,...,...,...,...,...,...
145,25,130,90,10.4,37.8,95.0,Low Risk
146,40,120,100,6.1,36.9,83.0,Low Risk
147,37,90,70,10.3,37.2,90.0,Low Risk
149,40,120,110,7.1,37.2,61.0,Low Risk


In [110]:
#checking HeartRate for unique values
data['HeartRate'].unique()

array([76.        , 82.        , 74.        , 96.        , 80.        ,
       73.        , 61.        , 70.        , 98.        , 97.        ,
       93.        , 77.        , 89.        , 86.        , 92.        ,
       83.        , 64.        , 88.        , 63.        , 69.        ,
       79.        , 65.        , 72.        , 75.        , 90.        ,
               nan, 67.        , 84.        , 62.        , 91.        ,
       80.76923077, 87.        , 81.        , 60.        , 71.        ,
       85.        , 68.        , 95.        , 94.        , 66.        ,
       99.        , 78.        ])

In [113]:
# sorting out with 250 HeartRate replacing for average
mean_HeartRate = data['HeartRate'].mean()
data.replace(250, mean_HeartRate, inplace=True)

In [114]:
# formating the data HeartRate
mean_HeartRate = data['HeartRate'].mean()
data['HeartRate'].fillna(mean_HeartRate, inplace=True)
data['HeartRate'] = round(data['HeartRate'], 1)
data['HeartRate'].unique()

/tmp/ipykernel_1704/1003010754.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['HeartRate'].fillna(mean_HeartRate, inplace=True)


array([76. , 82. , 74. , 96. , 80. , 73. , 61. , 70. , 98. , 97. , 93. ,
       77. , 89. , 86. , 92. , 83. , 64. , 88. , 63. , 69. , 79. , 65. ,
       72. , 75. , 90. , 79.6, 67. , 84. , 62. , 91. , 80.8, 87. , 81. ,
       60. , 71. , 85. , 68. , 95. , 94. , 66. , 99. , 78. ])

In [70]:
data['BodyTemp'] = Data_health_risk['BodyTemp']

In [71]:
#checking BodyTemp for unique values
data['BodyTemp'].unique()

array([ 98. ,  99. ,  36.6,   nan,  37.2, 100. ,  97.5])

In [89]:
# Anything above 50 is in degrees Fahrenheit. Anything below that is in degrees Celsius. I need Celsius
data.loc[data['BodyTemp'] > 50, 'BodyTemp'] = (data['BodyTemp'] - 32) / 1.8
data['BodyTemp'].unique()

array([36.66666667, 37.22222222, 36.6       ,         nan, 37.2       ,
       37.77777778, 36.38888889])

In [98]:
# formating the data BodyTemp
mean_BodyTemp = data['BodyTemp'].mean()
data['BodyTemp'].fillna(mean_BodyTemp, inplace=True)
data['BodyTemp'] = round(data['BodyTemp'], 1)
data['BodyTemp'].unique()

/tmp/ipykernel_1704/725184475.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['BodyTemp'].fillna(mean_BodyTemp, inplace=True)


array([36.7, 37.2, 36.6, 36.9, 37.8, 36.4])

In [48]:
#checking BS for unique values
data['BS'].unique()

array([10. ,  6.7,  6.2, 14.7, 13.5, 12.3,  9.7,  7.6,  7.4,  8.3, 10.9,
       12.4, 11.9,  8.5, 14.6, 12.6, 11. , 11.5,  9.8,  8.2,  nan,  6.1,
        7. ,  6.4, 13.7, 10.3,  6.9, 10.4,  9.9,  9.6, 11.7,  9.4, 11.6,
       10.5,  6.6, 11.8, 11.3, 14.5, 11.2,  9.5, 10.1, 14.1,  7.8,  6.8,
       12.1,  8.9, 13.6, 13.3,  7.1, 13.9, 13.2, 12.8, 14.9,  9.3, 13. ,
        9.1, 14.4,  0.1, 13.4,  8.6,  9. , 13.1,  6.5, 14. ,  9.2,  7.3,
       12.9, 14.2])

In [112]:
# sorting out with 0,1 BS replacing for average
mean_bs = data['BS'].mean()
data.replace(0.1, mean_bs, inplace=True)


In [109]:
data['BS'].fillna(mean_bs, inplace=True)
data['BS'] = round(data['BS'], 1) #round characters
data['BS'].unique()

/tmp/ipykernel_1704/495568630.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['BS'].fillna(mean_bs, inplace=True)


array([10. ,  6.7,  6.2, 14.7, 13.5, 12.3,  9.7,  7.6,  7.4,  8.3, 10.9,
       12.4, 11.9,  8.5, 14.6, 12.6, 11. , 11.5,  9.8,  8.2, 10.3,  6.1,
        7. ,  6.4, 13.7,  6.9, 10.4,  9.9,  9.6, 11.7,  9.4, 11.6, 10.5,
        6.6, 11.8, 11.3, 14.5, 11.2,  9.5, 10.1, 14.1,  7.8,  6.8, 12.1,
        8.9, 13.6, 13.3,  7.1, 13.9, 13.2, 12.8, 14.9,  9.3, 13. ,  9.1,
       14.4, 13.4,  8.6,  9. , 13.1,  6.5, 14. ,  9.2,  7.3, 12.9, 14.2])

In [50]:
#checking DiastolicBP for unique values
data['DiastolicBP'].unique()

array(['Eighty', '70', '100', '110', '60', 'Seventy', '90', '80', '85',
       nan, 'Sixty', 'Unknown'], dtype=object)

In [51]:
# replacement the incorrect data
data['DiastolicBP'] = data['DiastolicBP'].replace('Eighty', 80)
data['DiastolicBP'] = data['DiastolicBP'].replace('Seventy', 70)
data['DiastolicBP'] = data['DiastolicBP'].replace('Sixty', 60)
data['DiastolicBP'].unique()

array([80, '70', '100', '110', '60', 70, '90', '80', '85', nan, 60,
       'Unknown'], dtype=object)

In [101]:
#  Removing Unknown
data['DiastolicBP'] = data['DiastolicBP'].replace('Unknown', np.nan)
mean_dia = data['DiastolicBP'].mean()
data['DiastolicBP'].fillna(mean_dia, inplace=True)

/tmp/ipykernel_1704/3976319778.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['DiastolicBP'].fillna(mean_dia, inplace=True)


In [105]:
#  convertin characters (Casting)
data['DiastolicBP'] = data['DiastolicBP'].astype(int)

In [106]:
# checkin the result for column DiastolicBP
data['DiastolicBP'].unique()

array([ 80,  70, 100, 110,  60,  90,  85])

In [117]:
#checking SystolicBP for unique values
data['SystolicBP'].unique()

array([ nan, 160.,  90., 120., 110., 140., 130.])

In [121]:
#  Removing nan
mean_sys = data['SystolicBP'].mean()
data['SystolicBP'].fillna(mean_sys, inplace=True)
#  convertin characters to integer (Casting)
data['SystolicBP'] = data['SystolicBP'].astype(int)
data['SystolicBP'] = round(data['SystolicBP'] / 10) * 10 #round to the correct format
data['SystolicBP'] = data['SystolicBP'].astype(int)
data['SystolicBP'].unique()

/tmp/ipykernel_1704/3199626061.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['SystolicBP'].fillna(mean_sys, inplace=True)


array([120, 160,  90, 110, 140, 130])

In [56]:
#checking Age for unique values
data['Age'].unique()

array([ 52.,  45.,  33.,  nan,  29.,  18.,  60.,  35., 150.,  25.,  40.])

In [83]:
# sorting out with 150-years old guy
mean_age = data['Age'].mean()
data.replace(150, mean_age, inplace=True)
data['Age'].unique()

array([52.       , 45.       , 33.       ,        nan, 29.       ,
       18.       , 60.       , 35.       , 38.2519685, 25.       ,
       40.       ])

In [99]:
# formating the data column Age
mean_Age = data['Age'].mean()
data['Age'].fillna(mean_Age, inplace=True)      # and get a mean value (average) from the colum
data['Age'] = data['Age'].astype(int)           # CASTING TO INTEGER

/tmp/ipykernel_1704/3733144822.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Age'].fillna(mean_Age, inplace=True)      # and get a mean value (average) from the colum


In [100]:
#checking Age for unique values
data['Age'].unique()

array([52, 45, 33, 37, 29, 18, 60, 35, 38, 25, 40])

In [125]:
data

,Age,SystolicBP,DiastolicBP,BS,BodyTemp,HeartRate,RiskLevel
0,52,120,80,10.0,36.7,76.0,High Risk
1,45,160,70,6.7,36.7,82.0,High Risk
2,33,90,100,6.2,37.2,74.0,High Risk
3,37,120,110,14.7,36.6,96.0,High Risk
4,52,160,100,13.5,36.9,80.0,Low Risk
...,...,...,...,...,...,...,...
145,25,130,90,10.4,37.8,95.0,Low Risk
146,40,120,100,6.1,36.9,83.0,Low Risk
147,37,90,70,10.3,37.2,90.0,Low Risk
149,40,120,110,7.1,37.2,61.0,Low Risk


In [126]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 143 entries, 0 to 150
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Age          143 non-null    int64  
 1   SystolicBP   143 non-null    int64  
 2   DiastolicBP  143 non-null    int64  
 3   BS           143 non-null    float64
 4   BodyTemp     143 non-null    float64
 5   HeartRate    143 non-null    float64
 6   RiskLevel    143 non-null    object 
dtypes: float64(3), int64(3), object(1)
memory usage: 8.9+ KB


In [127]:
data.describe()

,Age,SystolicBP,DiastolicBP,BS,BodyTemp,HeartRate
count,143.000000,143.000000,143.000000,143.000000,143.000000,143.000000
mean,37.146853,123.776224,85.944056,10.316783,36.947552,80.074126
std,10.916429,19.565858,15.575250,2.625551,0.446083,10.803017
min,18.000000,90.000000,60.000000,6.100000,36.400000,60.000000
25%,29.000000,120.000000,70.000000,8.000000,36.600000,73.000000
50%,37.000000,120.000000,85.000000,10.300000,36.700000,80.000000
75%,45.000000,130.000000,100.000000,12.500000,37.200000,89.000000
max,60.000000,160.000000,110.000000,14.900000,37.800000,99.000000


Data has been cleaned

DATA ANALISYS (to be continue...)

In [128]:
print(data.groupby('RiskLevel')[['Age', 'BS']].mean())

                 Age         BS
RiskLevel                      
High Risk  37.300000  10.076667
Low Risk   38.623188  10.453623
mid risk   29.214286  10.671429
